### DISCUSSION TITLE
How to integrate Terra (planet lib) project with bevy

### DISCUSSION THREAD

**Discussion by greenmughal** [bevyengine/bevy#2210] [Q&A]
https://github.com/fintelia/terra
How can I plug terra into bevy, it requires access to webgpu device and queue atleast.

**Comment by bjorn3** [comment]
Bevy doesn't expose raw wgpu from it's api as it can work fine with other rendering api's too. For example bevy_webgl2 uses webgl2 to run bevy apps in the browser. Bevy is very flexible, so it is possible to completely avoid bevy_render/bevy_wgpu and instead let terra render everything I think. This means that you will have to manually render entities yourself. Alternatively you could fork bevy_wgpu to give you direct access to what you need. Just be careful that terra and bevy_wgpu don't run over each others by for example each clearing the screen as a first pass.

**Reply by greenmughal** [reply]
I had same hunch on that, thanks, looks like I need to fork both terra and bevy_wgpu.

**Reply by cart** [reply]
It is currently possible to get direct access to the wgpu device by retrieving the RenderResourceContext, downcasting to WgpuRenderResourceContext, then accessing the `Arc<Device>` in that struct. But the `Queue` is not currently exposed in userspace.

**Reply by cart** [reply]
https://github.com/bevyengine/bevy/blob/2fcac677129a451ecf03d369a7060e160c02a308/crates/bevy_wgpu/src/wgpu_renderer.rs#L86

**Comment by greenmughal** [comment]
Hi, I have forked both bevy and terra projects, I faced many issues but the somewhat good news is that both terra and bevy are initializing,
However to render terra planet I need to pass Depth and Color Texture/Buffer

```rust
pub fn terra_render(device: &wgpu::Device,
                queue: &wgpu::Queue,
                color_buffer: &wgpu::TextureView,
                depth_buffer: &wgpu::TextureView,
                size: (u32,u32),
                terrain: &mut Terrain)
```

But I dont know how to fight with render graph for these, the other way around might be to stuff terra_rendering code as a graph step or node?

If anyone is willing to help here, here is the modified codes of both projects

https://github.com/demon-hide/bevy
https://github.com/demon-hide/terra

It will atleast render without textures
```console
cargo run --example load_gltf            
```
